sawtooth-sdk-transaction-processor/src/main/java/sawtooth/sdk/processor/TransactionProcessor.java

```diff
@@ -187,57 +184,95 @@ private TransactionHandler findHandler(final Message message) {
       }
       LOGGER.info("Missing handler for header: " + header.toString());
     } catch (InvalidProtocolBufferException ipbe) {
-      LOGGER.info("Received Message that isn't a TransactionProcessRequest");
-      ipbe.printStackTrace();
+      LOGGER.log(Level.INFO, "Received Message that isn't a TransactionProcessRequest", ipbe);
     }
     return null;
   }
 
+  /**
+   * Whether or not this thread should keep running.
+   * @return true true if the processor should continue
+   */
+  private boolean keepRunning() {
+    synchronized (this) {
+      return keepRunningFlag;
+    }
+  }
+
+  /**
+   * Find a handler for this message and submit a new task for it. If no handler
+   * is found ignore and return.
+   * @param message message to process
+   */
+  private void submitTaskForMessage(final Message message) {
+    TransactionHandler handler = findHandler(message);
+    if (handler == null) {
+      return;
+    }
+    TransactionHandlerTask task = new TransactionHandlerTask(message, this.stream, handler);
+    executorService.submit(task);
+  }
+
+  /**
+   * Respond to a ping request.
+   * @param pingMessage incoming ping
+   */
+  private void respondToPing(final Message pingMessage) {
+    LOGGER.info("Recieved Ping Message.");
+    PingResponse pingResponse = PingResponse.newBuilder().build();
+    this.stream.sendBack(Message.MessageType.PING_RESPONSE, pingMessage.getCorrelationId(),
+        pingResponse.toByteString());
+  }
+
   @Override
   public final void run() {
-    while (true) {
+    while (keepRunning()) {
+      Message currentMessage;
       if (!this.handlers.isEmpty()) {
-        this.currentMessage = this.stream.receive();
-        if (this.currentMessage != null) {
-          if (this.currentMessage.getMessageType() == Message.MessageType.PING_REQUEST) {
-            LOGGER.info("Recieved Ping Message.");
-            PingResponse pingResponse = PingResponse.newBuilder().build();
-            this.stream.sendBack(Message.MessageType.PING_RESPONSE, this.currentMessage.getCorrelationId(),
-                pingResponse.toByteString());
-            this.currentMessage = null;
-          } else if (this.currentMessage.getMessageType() == Message.MessageType.TP_PROCESS_REQUEST) {
-            TransactionHandler handler = this.findHandler(this.currentMessage);
-            if (handler == null) {
-              break;
-            }
-            TransactionProcessor.process(this.currentMessage, this.stream, handler);
-            this.currentMessage = null;
+        currentMessage = this.stream.receive();
```

Comment:
Maybe check if running after the message is received in order to exit early (`receive` blocks)

Reply:
So looking closer at that there's a bigger problem under there.  first there is a null check in the run loop based on the receive(), but since the receive is based on LinkedBlockingQueue it can't ever return null, and nulls can't be added to it. So that code doesn't actually make sense.  Swapping it over to a receive(long) is more appropriate probably, but then there is a problem there to in that a no-result poll results in a TimeOutException.  In java that's not a good practice since a) it would likely be the more common result of the operation, and b) using exception processing as normal flow control is much much slower, exceptions should be used in exceptional cases.

Reply:
So, this has been refactored a bit in the the latest commit.